DOCS-437 update Rack-Awareness.md (#100)
aerospike/kubernetes-operator: docs/Rack-Awareness.md

description: Rack Awareness
---

In many situations, it's considered a "best practice" to build clusters which span multiple availability zones. Aerospike’s Rack Awareness feature lets you split your database across multiple racks or zones.

> For example, if you set a replication factor of 2, the master copy of the partition and its replica will be stored on separate hardware failure groups.

Reviewer: ```suggestion
For example, if you set a replication factor of 2, the master copy of the partition and its replica are stored on separate hardware failure groups.
```

> Rack Awareness also provides a mechanism which lets database clients read on a preferential basis from servers in their nearest rack or zone. This can provide lower latency, increase stability, and significantly reduce traffic charges by limiting cross-availability-zone traffic.

Reviewer: ```suggestion
Rack Awareness also provides a mechanism which lets database clients read from servers in their nearest rack or zone on a preferential basis. This can provide lower latency, increase stability, and significantly reduce traffic charges by limiting cross-availability-zone traffic.
```

          type: memory
```

Save and exit the CR file, then use kubectl to apply the change.

```shell
kubectl apply -f aerospike-cluster.yaml
```

For the full CR file, see the [example rack-enabled cluster CR](https://github.com/aerospike/aerospike-kubernetes-operator/blob/master/config/samples/rack_enabled_cluster_cr.yaml).

This and other example CRs are stored in [the main Aerospike Kubernetes Operator repository](https://github.com/aerospike/aerospike-kubernetes-operator/tree/master/config/samples).

## Cluster Node Distribution

> Cluster nodes are distributed across racks as evenly as possible. The cluster size is divided by the number of racks to get nodes per rack. Any remainder nodes are distributed one-by-one across racks starting from the first rack.

Reviewer: ```suggestion
Cluster nodes are distributed across racks as evenly as possible. The cluster size is divided by the number of racks to determine how many nodes per rack. Any remainder nodes are distributed one-by-one across racks starting from the first rack.
```

        zone: us-central1-c
```

Save and exit the CR file, then use kubectl to apply the change.

```shell
kubectl apply -f aerospike-cluster.yaml
```

> The Aerospike Kubernetes Operator redistributes cluster nodes across racks whenever the cluster size or the number of racks changes. If you add a rack without increasing the cluster size, the nodes will be redistributed. The number of nodes on existing racks will be scaled down, and the number of nodes on the new rack will be scaled up, per the usual topology rules.

Reviewer: ```suggestion
The Aerospike Kubernetes Operator redistributes cluster nodes across racks whenever the cluster size or the number of racks changes. If you add a rack without increasing the cluster size, the nodes are redistributed. The number of nodes on existing racks is scaled down, and the number of nodes on the new rack is scaled up, per the usual topology rules.
```

## Set Rack-Level Storage and aerospikeConfig

> Aerospike's Rack Awareness lets you set local storage and aerospikeConfig options. If you provide local storage for a rack, the rack will use this storage. Otherwise, common global storage will be used.

Reviewer: ```suggestion
Aerospike's Rack Awareness lets you set local storage and aerospikeConfig options. If you provide local storage for a rack, the rack uses this storage. Otherwise, common global storage is used.
```

> In the following example, the aerospikeConfig is a patch which will be used for the rack and merged with the common global aerospikeConfig.

Reviewer: ```suggestion
In the following example, the aerospikeConfig is a patch which is used for the rack, and merged with the common global aerospikeConfig.
```

                path: /etc/aerospike/secret
```

Save and exit the CR file, then use kubectl to apply the change.

```shell
kubectl apply -f aerospike-cluster.yaml
```

## Merge AerospikeConfig

> A rack's local aerospikeConfig patch will be merged with the common global base aerospikeConfig based on the following rules:

Reviewer: ```suggestion
A rack's local aerospikeConfig patch is merged with the common global base aerospikeConfig based on the following rules:
```

> * New elements from the patch configMap will be added to the base configMap.

Reviewer: ```suggestion
* New elements from the patch configMap are added to the base configMap.
```

> * A base element will be replaced with a new patch element if:

Reviewer: ```suggestion
* A base element is replaced with a new patch element if:
```

    * The element value type is changed.
    * The element value is a primitive type and updated.
    * The element value is a primitive list type and updated.
    * The element key is `storage-engine` and the storage-engine type has been changed (storage-engine can be of `device`, `file`, or `memory` type).

> * If the element is of map type, patch and base elements will be recursively merged.

Reviewer: ```suggestion
* If the element is of map type, patch and base elements is recursively merged.
```

> * If the elements are list of maps, new list elements in the patch list will be appended to the base list and corresponding entries will be merged using the same merge algorithm.

Reviewer: ```suggestion
* If the elements are list of maps, new list elements in the patch list is appended to the base list and corresponding entries are merged using the same merge algorithm.
```

>     * The order of elements in the base list will be maintained.

Reviewer: ```suggestion
    * The order of elements in the base list is maintained.
```

          type: memory
```

Save and exit the CR file, then use kubectl to apply the change.

```shell
kubectl apply -f aerospike-cluster.yaml
```

## Remove a Rack

To remove a rack, delete that rack's section from the `rackConfig.racks` section, then use kubectl to apply the new configuration.

> The Aerospike Kubernetes Operator will scale down the desired rack to size 0 by removing one node at a time from the rack. After all the nodes have been removed, the rack will also be removed. If you are removing a rack without decreasing cluster size, other racks will be scaled up based on new node redistribution.

Reviewer: ```suggestion
The Aerospike Kubernetes Operator scales down the desired rack to size 0 by removing one node at a time from the rack. After all the nodes have been removed, the rack is removed. If you are removing a rack without decreasing cluster size, other racks are scaled up based on new node redistribution.
```
